### PR TITLE
`Development`: Migrate the client modeling module to signals and vitest

### PR DESCRIPTION
**Pull request by AjayvirS**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I **strictly** followed the [client coding guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).

### Motivation and Context
This PR migrates the modeling module to Angular 21 best practices, specifically:
- Converting legacy `@Input()` and `@Output()` decorators to signal-based `input()` and `output()` APIs
- Migrating `@ViewChild` decorators to `viewChild()` signals
- Migrating tests from Jest to Vitest with zoneless testing
- Replacing `ngOnChanges` lifecycle hooks with `effect()` for reactive data flow
 Description
**Signal Migrations:**
- Converted all `@Input()` decorators to `input()` signals across modeling components
- Converted all `@Output()` decorators to `output()` signals
- Migrated `@ViewChild` references to `viewChild()` signals in:
  - `modeling.component.ts` (base class)
  - `modeling-submission.component.ts`
  - `example-modeling-submission.component.ts`
  - `modeling-exercise-update.component.ts`
- Used `model()` signal for two-way bound `explanation` field
**Test Migration (Jest → Vitest):**
- Migrated all 15 modeling spec files to Vitest with `setupTestBed({ zoneless: true })`
- Replaced `jest.fn()` / `jest.spyOn()` with `vi.fn()` / `vi.spyOn()`
- Updated test patterns to use `fixture.componentRef.setInput()` for signal inputs
- Created new `fullscreen.component.spec.ts` (19 tests) for previously untested component
- Added modeling module to `vitest.config.ts` test includes and coverage

#### Prerequisites
- 1 Course
- 1 Instructor
- 1 Student
- 1 **Modeling Exercise** with:
  - Automatic assessment disabled
- The student has created and submitted at least one modeling submission for this exercise

#### Flow 1 – Regular modeling submission & assessment

1. Log in as **Student**.
2. Navigate to the course and open the **Modeling Exercise**.
3. Open the online modeling editor, make a small change to the model (add/move an element), optionally edit the explanation text, and **submit** the exercise.
4. Log out and log in as **Tutor**.
5. Go to  
   `Course Management` &rarr; `[Course]` &rarr; `Assessment Dashboard` &rarr; `[Modeling Exercise]` &rarr; `Exercise Dashboard` &rarr; `Start Manual Assessment` &rarr; `Start new assessment`.
7. Verify:
   - The model is displayed correctly in the modeling editor.
   - The score header / score display shows max points and current total score.
   - The editor is interactive (you can select diagram elements with a blue outline for selected element).
8. Assess the submission:
    - Add at least two referenced feedbacks by double-clicking diagram elements and entering credits + text.
   - (Optional) Add one **unreferenced** feedback in the general feedback area.
   - Save the assessment.
9. Reload the assessment page (or navigate away and back to the submission) and verify:
   - All feedback is still present and attached to the correct elements.
   - The total score matches the sum of the feedback credits.
   - No console errors occur.
10. Login as the student who made the submission and verify that the referenced feedback and unreferenced feedback is shown correctly to the student

#### Flow 2 – Example modeling submission & assessment training

1. Log in as **Instructor**
2. Navigate to  
   `Course Management &rarr; [Course] &rarr; Exercises &rarr; [Modeling Exercise] &rarr; `Example Submissions` &rarr; `+ Create Example Submission`.
3. Create an**example modeling submission** with two elements with `Assess Correctly` toggled and press on `Create new Example Submission`.
4. Press on `Show Assessment`.
5. In the **modeling editor** view:
   - Verify the example model is rendered.
   - Add feedback with >0 points to one element and <0 points to another.
   - Save the example assessment.
   - Reload the page and confirm that the model with feedback is persisted.
6. Navigate to the `Assessment` dashboard view and press on `Exercise Dashboard` for the relevant exercise:
   - Press on the `Start assessing example submissions` button under `Assessed submissions` header.
   - Give the same referenced feedback as you did for the example feedback and submit assessment.
   - Confirm that you get a `Congratulations! This is a good assessment!` message.
   - 
   - Adjust the feedback to make a wrong assessment
   - Verify that the modeling view is updated correctly with the legend for `Incorrect assessment`

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Lines | Expects | Ratio |
|------------|-------------:|------:|--------:|------:|
| example-submission-import.component.ts | 96.42% | 59 | 3 | 5.1 |
| modeling-assessment.component.ts | 91.13% | 285 | 58 | 20.4 |
| modeling-assessment.util.ts | 89.61% | 119 | ? | ? |
| example-modeling-submission.component.ts | 90.04% | 445 | 63 | 14.2 |
| modeling-exercise.component.ts | 94.59% | 91 | 18 | 19.8 |
| modeling-exercise-update.component.ts | 85.34% | 306 | 43 | 14.1 |
| modeling-submission.component.ts | 85.51% | 676 | 83 | 12.3 |
| fullscreen.component.ts | 100.00% | 29 | 23 | 79.3 |
| modeling-editor.component.ts | 86.30% | 157 | 26 | 16.6 |
| modeling-explanation-editor.component.ts | 100.00% | 17 | 7 | 41.2 |
| modeling.component.ts | 57.14% | 55 | ? | ? |

## Summary by CodeRabbit

* **Refactor**
  * Modernized components and templates to use a consistent reactive binding approach across modeling editors and views, improving responsiveness and maintainability.

* **Bug Fixes**
  * Improved stability for assessments, feedback syncing, highlighting, resizing, fullscreen and print rendering; avoids errors when model or feedback data is missing and ensures correct feedback/display in example and team views.

* **Tests**
  * Updated tests to provide inputs via binding mechanisms and validate the new reactive behaviors and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts [hunk 1]
```diff
@@ -27,26 +27,21 @@ export interface DropInfo {
...
-    @Input() maxBonusPoints = 0;
-    @Input() totalScore: number;
-    @Input() title: string;
-    @Input() enablePopups = true;
-    @Input() displayPoints = true;
-    @Input() highlightDifferences: boolean;
+    maxScore = input<number>(undefined!);
+    maxBonusPoints = input(0);
+    totalScore = input<number>(undefined!);
+    readonly title = input<string | undefined>(undefined);
```
Comment (konstantinkissel) [inline src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: Why is the title readonly but everything else isn't?

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts [hunk 2]
```diff
@@ -27,26 +27,21 @@ export interface DropInfo {
...
-    @Input() maxScore: number;
-    @Input() maxBonusPoints = 0;
-    @Input() totalScore: number;
-    @Input() title: string;
-    @Input() enablePopups = true;
-    @Input() displayPoints = true;
-    @Input() highlightDifferences: boolean;
+    maxScore = input<number>(undefined!);
+    maxBonusPoints = input(0);
+    totalScore = input<number>(undefined!);
```
Comment (konstantinkissel) [inline src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: Why are we switching between using `undefined!` and `undefined`?
To me `undefined!` seems like a workaround to trick TypeScript

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts [hunk 3]
```diff
@@ -27,26 +27,21 @@ export interface DropInfo {
...
+    readonly title = input<string | undefined>(undefined);
+    enablePopups = input(true);
+    displayPoints = input(true);
+    highlightDifferences = input<boolean | undefined>(undefined);
```
Comment (konstantinkissel) [inline src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: Since highlightDifferences is not marked as required, won't `highlightDifferences = input<boolean>();` achieve the same result?

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts [hunk 4]
```diff
@@ -475,17 +475,23 @@ export class ExampleModelingSubmissionComponent implements OnInit, FeedbackMarke
...
+            if (validatedFeedback) {
                 validatedFeedback.correctionStatus = correctionError.type;
             }
         });
-        this.assessmentEditor.resultFeedbacks = this.referencedFeedback;
+
+        // we reassign to trigger change detection
+        this.referencedFeedback = [...this.referencedFeedback];
+        this.unreferencedFeedback = [...this.unreferencedFeedback];
+
```
Comment (konstantinkissel) [inline src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts]: We should definitely rework this. Creating shallow copies of the arrays to trigger updates defeats the purpose of a signal migration, which is meant to make change detection more efficient.

We should consider turning `referencedFeedback` and `unreferencedFeedback` into signals.

File: src/main/webapp/app/modeling/overview/modeling-submission/modeling-submission.component.ts [hunk 5]
```diff
@@ -171,19 +171,25 @@ export class ModelingSubmissionComponent implements OnInit, OnDestroy, Component
     isFeedbackView = false;
     showResultHistory = false;
 
+    // since participationId can come from parent component (readonly signal) or from route, we use an internal writeable
+    // variable
+    private effectiveParticipationId?: number;
```
Comment (konstantinkissel) [inline src/main/webapp/app/modeling/overview/modeling-submission/modeling-submission.component.ts]: We should ensure that the `effectiveParticipationId` can respond to changes, otherwise it's just a workaround. A better approach would be to turn it into a computed variable, e.g. like this:
```
private routeParams = toSignal(this.route.params);
effectiveParticipationId = computed(() => {
        return this.participationId() ?? Number(this.routeParams()?.['participationId']);
    });
```

File: src/main/webapp/app/modeling/shared/modeling/modeling.component.ts [hunk 6]
```diff
@@ -13,19 +13,23 @@ export abstract class ModelingComponent {
...
-    @Input() diagramType?: UMLDiagramType;
-    @Input() explanation: string;
-    @Input() readOnly = false;
+    resizeOptions = input<{
+        horizontalResize?: boolean;
+        verticalResize?: boolean;
+    }>(undefined!);
+    umlModel = input<UMLModel>(undefined!);
+    diagramType = input<UMLDiagramType>();
+    explanation = model<string>(undefined!);
```
Comment (konstantinkissel) [inline src/main/webapp/app/modeling/shared/modeling/modeling.component.ts]: Here again we use `undefined!`, which is a workaround.

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts (hunk 1 above)
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: Thank you for the code review! None of them are supposed to be readonly since the parent components assign values to it, fixed!

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts (hunk 2 above)
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: Looks like the automigration tool explicitly sets undefined, I have manually changed them and set sensible default values

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts (hunk 3 above)
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: fixed!

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts [hunk 7]
```diff
@@ -57,6 +52,29 @@ export class ModelingAssessmentComponent extends ModelingComponent implements Af
...
+            const incoming = this.resultFeedbacks();
+
+            if (!incoming) {
+                return;
+            }
+
+            this.feedbacks = incoming;
+            this.referencedFeedbacks = this.feedbacks.filter((feedbackElement) => feedbackElement.reference != undefined);
+            this.updateApollonAssessments(this.referencedFeedbacks);
+        });
```
[the remark this answers is not in the document]
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts:76]: unreferenced feedbacks does not need to be filtered

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts [hunk 8]
```diff
@@ -24,29 +24,24 @@ export interface DropInfo {
...
-    @Input() set resultFeedbacks(feedback: Feedback[]) {
-        this.feedbacks = feedback;
-        this.referencedFeedbacks = this.feedbacks.filter((feedbackElement) => feedbackElement.reference != undefined);
-        this.updateApollonAssessments(this.referencedFeedbacks);
-    }
+    highlightedElements = input<Map<string, string> | undefined>(undefined); // map elementId -> highlight color
+    elementCounts = input<ModelElementCount[]>();
+    course = input<Course>();
 
     feedbacks: Feedback[];
```
Comment (Elfari1028) [inline src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: since `set resultFeedbacks` would simply set `this.feedbacks = feedback;`, we probably can remove this `feedback` field altogether? and use values from `resultFeedbacks` instead?

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts [hunk 9]
```diff
@@ -57,6 +52,41 @@ export class ModelingAssessmentComponent extends ModelingComponent implements Af
...
+        });
+        effect(() => {
+            const incoming = this.resultFeedbacks();
+
+            if (!incoming) {
+                return;
+            }
+
+            this.feedbacks = incoming;
+            this.referencedFeedbacks = this.feedbacks.filter((feedbackElement) => feedbackElement.reference != undefined);
```
Comment (Elfari1028) [inline src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts]: same here, 
```
this.referencedFeedbacks = this.resultFeedbacks().filter((feedbackElement) => feedbackElement.reference != undefined);
``` would suffice

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.html [hunk 10]
```diff
@@ -185,7 +186,7 @@ <h3 class="mt-3">
                     @if (assessmentMode) {
                         @if (result && result.id) {
                             <jhi-unreferenced-feedback
-                                [(feedbacks)]="unreferencedFeedback"
```
Comment (Elfari1028) [inline src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.html:188]: well done, i'm not sure why originally it is done as two way binding even tho `(feedbacksChange)="onUnReferencedFeedbackChanged($event)"` existed

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts [hunk 11]
```diff
@@ -114,13 +114,10 @@ export class ExampleModelingSubmissionComponent implements OnInit, FeedbackMarke
     ];
 
     private exampleSubmissionId: number;
+    referencedFeedback = signal<Feedback[]>([]);
```
Comment (Elfari1028) [inline src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts:115]: nicely done here by turning these into signals

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts [hunk 12]
```diff
@@ -465,33 +462,40 @@ export class ExampleModelingSubmissionComponent implements OnInit, FeedbackMarke
...
-        getLatestSubmissionResult(exampleSubmission.submission)!.feedbacks = this.assessments;
+        getLatestSubmissionResult(exampleSubmission.submission)!.feedbacks = this.assessments();
 
         const command = new ExampleSubmissionAssessCommand(this.tutorParticipationService, this.alertService, this);
         command.assessExampleSubmission(exampleSubmission, this.exerciseId);
     }
 
     markAllFeedbackToCorrect() {
-        this.assessments.forEach((feedback) => {
+        this.assessments().forEach((feedback) => {
```
Comment (Elfari1028) [inline src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts]: assessment is computed, why are we interating over each feedback to mark it as correct?

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts [hunk 13]
```diff
@@ -465,33 +462,40 @@ export class ExampleModelingSubmissionComponent implements OnInit, FeedbackMarke
...
 
     markWrongFeedback(correctionErrors: FeedbackCorrectionError[]) {
-        correctionErrors.forEach((correctionError) => {
-            const validatedFeedback = this.assessments.find((feedback) => feedback.reference === correctionError.reference);
-            if (validatedFeedback != undefined) {
-                validatedFeedback.correctionStatus = correctionError.type;
+        const byReference = new Map(correctionErrors.map((err) => [err.reference, err]));
+
+        // mutate referenced feedback
+        const referenced = this.referencedFeedback();
```
Comment (Elfari1028) [inline src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts:479]: in old code,we are using `this.assessments`, and here we are using `referencedFeedback` and not `assessments`. why is that?

now we are only iterating over `this.referencedFeedback();` and not `asssesments` .. not sure if i understand this change.

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts (hunk 12 above)
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts]: You are right, since we update the referencedFeedback and unreferencedFeedback anyway, assessments() is just a computed value and it should get updated through the other 2 signals.

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts (hunk 13 above)
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts:479]: Before refactor we had to set `this.assessmentEditor.resultFeedbacks = this.referencedFeedback;` for `modeling-assessment.component.ts`, however after signal migration `referencedFeedback` is a signal now and passed in the template itself so we do not need to iterate over the computed assessments anymore:

This was also confusing for me. Only referenced feedback is considered for correction errors. In `modeling-assessment.component.ts` we do not really do anything with `unreferencedFeedbacks` apart from apart from filtering it. However I kept that behavior in, incase other components which use `modeling-assessment.component.ts` need it in the future. 

From a user perspective, I think it makes sense that unreferencedFeedback does not need to be assessed because for example submissions we only evaluate referenced feedback. However if you think it makes sense to get rid of unreferencedFeedback in `modeling-assessment.component.ts`, we can do that, let me know what you think!

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts [hunk 14]
```diff
@@ -465,33 +462,36 @@ export class ExampleModelingSubmissionComponent implements OnInit, FeedbackMarke
...
+        referenced.forEach((feedback) => {
+            const err = byReference.get(feedback.reference!);
+            if (err) {
+                feedback.correctionStatus = err.type;
             }
         });
-        this.assessmentEditor.resultFeedbacks = this.referencedFeedback;
+        this.referencedFeedback.set(referenced);
+
```
[the remark this answers is not in the document]
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts]: this changes older behavior semantics and would break tests, it should not be changed unless the test specification changes

File: src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts [hunk 15]
```diff
@@ -24,31 +24,25 @@ export interface DropInfo {
...
+    displayPoints = input(true);
+    highlightDifferences = input<boolean>();
+    resultFeedbacks = input<Feedback[]>();
+
+    feedbackChanged = output<Feedback[]>();
+    selectionChanged = output<Selection>();
+
+    highlightedElements = input<Map<string, string> | undefined>(undefined); // map elementId -> highlight color
+    elementCounts = input<ModelElementCount[]>();
+    course = input<Course>();
```
[the remark this answers is not in the document]
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/assess/modeling-assessment.component.ts:45]: this changes older behavior semantics and would break tests, it should not be changed unless the test specification changes

File: src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (AjayvirS) [reply src/main/webapp/app/modeling/manage/example-modeling/example-modeling-submission.component.ts:487]: This is not completely true for Angular signals, calling `.set(..)` does not break reactivity, even if the reference to the object in question remains the same and we need mutability here.

**Review by KonstiAnon — DISMISSED** [review]
Code

**Review by jerrycai0006 — DISMISSED** [review]
Tested on TS4, works as expected.

**Comment by matyasht** [comment]
In Flow 2 I encountered some issues when testing on TS2 and locally:
1. When I type in the fields as you see in the video they instantly close.
2. When I give feedback in an 0 point field I can't submit the assesment anymore.

Video: https://youtu.be/wbAT3yyIlI4

File: src/main/webapp/app/modeling/manage/update/modeling-exercise-update.component.spec.ts [hunk 16]
```diff
@@ -1,367 +1,559 @@
...
+    }
+}
+
+// Stub for MarkdownEditorMonacoComponent
+@Component({ selector: 'jhi-markdown-editor-monaco', template: '' })
+class StubMarkdownEditorMonacoComponent {
+    @Input() markdown: string = '';
+    @Input() domainActions: unknown[] = [];
+    markdownChange = output<string>();
+}
```
Comment (FelixTJDietrich) [inline src/main/webapp/app/modeling/manage/update/modeling-exercise-update.component.spec.ts:115]: We should use signals here?

**Comment by AjayvirS** [comment]
> In Flow 2 I encountered some issues when testing on TS2 and locally:
> 
>     1. When I type in the fields as you see in the video they instantly close.
> 
>     2. When I give feedback in an 0 point field I can't submit the assesment anymore.
> 
> 
> Video: https://youtu.be/wbAT3yyIlI4

Thank you for testing and noticing this bug, it happened due to an incorrect signal being passed to a child component which led to the editor resetting and the dialogue pane closing, I have reverted that change to the way it was before.

File: src/main/webapp/app/modeling/manage/update/modeling-exercise-update.component.spec.ts (hunk 16 above)
Comment (krusche) [reply src/main/webapp/app/modeling/manage/update/modeling-exercise-update.component.spec.ts:115]: yes

**Review by KonstiAnon — DISMISSED** [review]
Code

**Review by FelixTJDietrich — APPROVED** [review]
Code looks good, I tested extensively on TS2 and fixed an exam mode edit exercise issue. Everything works as expected for me now after retesting.
